Setting dbms.security.procedures.roles to empty string causes ArrayIndexOutOfBoundsException

Setting config parameter dbms.security.procedures.roles to an empty string, causes a failing database start with java.lang.ArrayIndexOutOfBoundsException.

**Neo4j Version:** 3.1.3
**Operating System:** Centos 6.8
**API:** -

### Steps to reproduce
1. Setup vanilla neo4j with version 3.1.3
2. Add the following line to conf/neo4j.conf:
`dbms.security.procedures.roles=`
3. Start the database with bin/neo4j start

### Expected behavior
Data should startup normally.

### Actual behavior
Startup crashes with the following stack trace:
```java
2017-04-11 13:03:48.640+0000 WARN  WARNING! Deprecated configuration options used. See manual for details
2017-04-11 13:03:48.641+0000 WARN  dbms.index_sampling.buffer_size has been replaced with dbms.index_sampling.sample_size_limit.
2017-04-11 13:03:48.675+0000 INFO  ======== Neo4j 3.1.3 ========
2017-04-11 13:03:48.889+0000 INFO  Starting...
2017-04-11 13:03:49.032+0000 ERROR Failed to start Neo4j: Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabase@38c6f217' was successfully initialized, but failed to start. Please see attached cause exception. Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabase@38c6f217' was successfully initialized, but failed to start. Please see attached cause exception.
org.neo4j.server.ServerStartupException: Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabase@38c6f217' was successfully initialized, but failed to start. Please see attached cause exception.
	at org.neo4j.server.exception.ServerStartupErrors.translateToServerStartupError(ServerStartupErrors.java:68)
	at org.neo4j.server.AbstractNeoServer.start(AbstractNeoServer.java:230)
	at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:91)
	at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:68)
	at org.neo4j.server.CommunityEntryPoint.main(CommunityEntryPoint.java:28)
Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.server.database.LifecycleManagingDatabase@38c6f217' was successfully initialized, but failed to start. Please see attached cause exception.
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:443)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
	at org.neo4j.server.AbstractNeoServer.start(AbstractNeoServer.java:202)
	... 3 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: 1
	at org.neo4j.kernel.impl.proc.ProcedureAllowedConfig.lambda$new$1(ProcedureAllowedConfig.java:61)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at org.neo4j.kernel.impl.proc.ProcedureAllowedConfig.<init>(ProcedureAllowedConfig.java:64)
	at org.neo4j.kernel.impl.factory.DataSourceModule.setupProcedures(DataSourceModule.java:356)
	at org.neo4j.kernel.impl.factory.DataSourceModule.<init>(DataSourceModule.java:176)
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.createDataSource(GraphDatabaseFacadeFactory.java:243)
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.initFacade(GraphDatabaseFacadeFactory.java:151)
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.newFacade(GraphDatabaseFacadeFactory.java:130)
	at org.neo4j.server.CommunityNeoServer.lambda$static$0(CommunityNeoServer.java:57)
	at org.neo4j.server.database.LifecycleManagingDatabase.start(LifecycleManagingDatabase.java:89)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:433)
	... 5 more
```

--- a/community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureAllowedConfig.java
+++ b/community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureAllowedConfig.java
@@ -52,9 +52,10 @@ public class ProcedureAllowedConfig
     {
         Map<String,String> params = config.getParams();
         this.defaultValue = params.get( PROC_ALLOWED_SETTING_DEFAULT_NAME );
-        if ( params.containsKey( PROC_ALLOWED_SETTING_ROLES ) )
+        String allowedRoles = params.get( PROC_ALLOWED_SETTING_ROLES );
+        if ( allowedRoles != null && allowedRoles.length() != 0 )
         {
-            this.matchers = Stream.of( params.get( PROC_ALLOWED_SETTING_ROLES ).split( SETTING_DELIMITER ) )
+            this.matchers = Stream.of( allowedRoles.split( SETTING_DELIMITER ) )
                     .map( procToRoleSpec ->
                     {
                         String[] spec = procToRoleSpec.split( MAPPING_DELIMITER );

--- a/community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureAllowedConfig.java
+++ b/community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureAllowedConfig.java
@@ -52,13 +52,12 @@ public class ProcedureAllowedConfig
     {
         Map<String,String> params = config.getParams();
         this.defaultValue = params.get( PROC_ALLOWED_SETTING_DEFAULT_NAME );
-        String allowedRoles = params.get( PROC_ALLOWED_SETTING_ROLES );
-        if ( allowedRoles != null && allowedRoles.length() != 0 )
+        if ( params.containsKey( PROC_ALLOWED_SETTING_ROLES ))
         {
-            this.matchers = Stream.of( allowedRoles.split( SETTING_DELIMITER ) )
-                    .map( procToRoleSpec ->
+            this.matchers = Stream.of( params.get( PROC_ALLOWED_SETTING_ROLES ).split( SETTING_DELIMITER ) )
+                    .map( procToRoleSpec -> procToRoleSpec.split( MAPPING_DELIMITER ) )
+                    .filter( spec -> spec.length > 1 ).map( spec ->
                     {
-                        String[] spec = procToRoleSpec.split( MAPPING_DELIMITER );
                         String[] roles = stream( spec[1].split( ROLES_DELIMITER ) )
                                 .map( String::trim ).toArray( String[]::new );
                         return new ProcMatcher( spec[0].trim(), roles );

--- a/community/kernel/src/test/java/org/neo4j/kernel/impl/proc/ProcedureAllowedConfigTest.java
+++ b/community/kernel/src/test/java/org/neo4j/kernel/impl/proc/ProcedureAllowedConfigTest.java
@@ -67,6 +67,29 @@ public class ProcedureAllowedConfigTest
     }
 
     @Test
+    public void shouldNotFailOnEmptyStringDefaultName()
+    {
+        Config config = Config.defaults()
+                .with( genericMap( PROC_ALLOWED_SETTING_DEFAULT_NAME, ""));
+        new ProcedureAllowedConfig( config );
+    }
+    @Test
+    public void shouldNotFailOnEmptyStringRoles()
+    {
+        Config config = Config.defaults()
+                .with( genericMap( PROC_ALLOWED_SETTING_ROLES, "" ) );
+        new ProcedureAllowedConfig( config );
+    }
+    @Test
+    public void shouldNotFailOnEmptyStringBoth()
+    {
+        Config config = Config.defaults()
+                .with( genericMap( PROC_ALLOWED_SETTING_DEFAULT_NAME, "",
+                        PROC_ALLOWED_SETTING_ROLES, "" ) );
+       new ProcedureAllowedConfig( config );
+    }
+
+    @Test
     public void shouldHaveConfigsWithWildcardProcedureAllowed()
     {
         Config config = Config.defaults()

--- a/community/kernel/src/test/java/org/neo4j/kernel/impl/proc/ProcedureAllowedConfigTest.java
+++ b/community/kernel/src/test/java/org/neo4j/kernel/impl/proc/ProcedureAllowedConfigTest.java
@@ -73,6 +73,7 @@ public class ProcedureAllowedConfigTest
                 .with( genericMap( PROC_ALLOWED_SETTING_DEFAULT_NAME, ""));
         new ProcedureAllowedConfig( config );
     }
+
     @Test
     public void shouldNotFailOnEmptyStringRoles()
     {
@@ -80,6 +81,15 @@ public class ProcedureAllowedConfigTest
                 .with( genericMap( PROC_ALLOWED_SETTING_ROLES, "" ) );
         new ProcedureAllowedConfig( config );
     }
+
+    @Test
+    public void shouldNotFailOnBadStringRoles()
+    {
+        Config config = Config.defaults()
+                .with( genericMap( PROC_ALLOWED_SETTING_ROLES, "matrix" ) );
+        new ProcedureAllowedConfig( config );
+    }
+
     @Test
     public void shouldNotFailOnEmptyStringBoth()
     {

